daemon/info.go

```diff
@@ -233,23 +239,48 @@ func memInfo() *system.MemInfo {
 	return memInfo
 }
 
+// this caches the result of operatingSystem below, for efficiency
+var operatingSystemResult string
+
 func operatingSystem() string {
-	var operatingSystem string
-	if s, err := operatingsystem.GetOperatingSystem(); err != nil {
-		logrus.Warnf("Could not get operating system name: %v", err)
-	} else {
-		operatingSystem = s
+	defer metrics.StartTimer(hostInfoFunctions.WithValues("operating_system"))()
+
+	if operatingSystemResult == "" {
+		var operatingSystem string
+		if s, err := operatingsystem.GetOperatingSystem(); err != nil {
+			logrus.Warnf("Could not get operating system name: %v", err)
+		} else {
+			operatingSystem = s
+		}
+		// Don't do containerized check on Windows
+		if runtime.GOOS != "windows" {
+			if inContainer, err := operatingsystem.IsContainerized(); err != nil {
+				logrus.Errorf("Could not determine if daemon is containerized: %v", err)
+				operatingSystem += " (error determining if containerized)"
+			} else if inContainer {
+				operatingSystem += " (containerized)"
+			}
+		}
+
+		operatingSystemResult = operatingSystem
 	}
-	// Don't do containerized check on Windows
-	if runtime.GOOS != "windows" {
-		if inContainer, err := operatingsystem.IsContainerized(); err != nil {
-			logrus.Errorf("Could not determine if daemon is containerized: %v", err)
-			operatingSystem += " (error determining if containerized)"
-		} else if inContainer {
-			operatingSystem += " (containerized)"
+	return operatingSystemResult
+}
+
+// this caches the result of osVersion below, for efficiency
+var osVersionResult string
+
+func osVersion() string {
+	defer metrics.StartTimer(hostInfoFunctions.WithValues("os_version"))()
+
+	if osVersionResult == "" {
```

Comment:
Same here; could use an early return;

```suggestion
	if osVersionResult != "" {
		return osVersionResult
	}
```